safe-logging/src/main/java/com/palantir/logsafe/Arg.java

```diff
@@ -18,16 +18,20 @@
 
 import java.io.Serializable;
 import java.util.Objects;
+import java.util.function.Supplier;
 
 /** A wrapper around an argument used to build a formatted message. */
 public abstract class Arg<T> implements Serializable {
 
     private final String name;
-    private final T value;
+    private final Supplier<T> lazyValue;
 
-    protected Arg(String name, T value) {
+    private boolean cached;
+    private T value;
```

Comment:
See below, but I don't think we want to eat these extra fields for non-lazy args